Feature/dissipative controller

buschbapti opened the pull request:
Here it is, PassiveDS as an Impedance controller. Simple test case, we should do a proper comparison with already existing code.

TODO:
- the notion of energy tank is not in there yet

eeberhard commented on source/controllers/tests/testDissipativeImpedanceController.cpp:
```diff
@@ -0,0 +1,106 @@
+  damping = controller_.get_damping();
+
+  // linear part is identity
+  for (int i = 0; i < 3; ++i) {
```
Rather than indexed for loops to check for zero values, you could use either of the follow approaches:

```c++
#include <numeric>
auto block = damping.block<3,3>(0, 0);
EXPECT_NEAR(std::accumulate(block.data(), block.data()+block.size(), 0.0), 0.0, tolerance)
```

```c++
for(auto elem : damping.block<3,3>(0, 0).reshaped()) {
  EXPECT_NEAR(elem, block.data()+block.size(), 0.0), 0.0, tolerance)
}
```

buschbapti replied:
neat thanks

eeberhard commented on source/controllers/include/controllers/impedance/Impedance.hpp:
```diff
@@ -35,7 +35,7 @@ class Impedance : Controller<S, S> {
    * @return the output command at the input state
    */
-  virtual S compute_command(const S& desired_state, const S& feedback_state) const;
+  virtual S compute_command(const S& desired_state, const S& feedback_state);
```
Ah yes I noticed this right at the end of the previous PR, it was therefore hiding the base definition. Glad you caught it here.

eeberhard commented on source/controllers/tests/testDissipativeImpedanceController.cpp:
```diff
@@ -0,0 +1,106 @@
+    for (int j = 0; j < 6; ++j) { EXPECT_NEAR(damping(i, j), identity6(i, j), 10e-4); }
+  }
+}
+
```
I suggest you add a test case that sets a different damping value in the principle direction than the secondary ones, and then check the output of the command with a reference velocity parallel and orthogonal to the input velocity. Check that the output is proportional to the first and second damping values accordingly

eeberhard reviewed (approved):
Looks good to me apart from minor comments. The math seems to correspond with what I've seen in the other controller. I also appreciate the use of test fixtures in this latest test code!

buschbapti commented on source/controllers/src/impedance/Dissipative.cpp:
```diff
@@ -0,0 +1,59 @@
+
+void Dissipative::compute_damping(const Eigen::VectorXd& desired_twist) {
+  switch (this->computational_space_) {
+    case ComputationalSpaceType::LINEAR:
```
of course this is so silly thanks

buschbapti commented on source/controllers/tests/testDissipativeImpedanceController.cpp:
```diff
@@ -0,0 +1,106 @@
+#include "controllers/impedance/Dissipative.hpp"
+#include <gtest/gtest.h>
+
+using namespace controllers::impedance;
```
yes I agree I tend to avoid using namespace actually but the code becomes rapidly verbose without.

buschbapti commented on source/controllers/tests/testDissipativeImpedanceController.cpp:
```diff
@@ -0,0 +1,106 @@
+  Eigen::Matrix3d identity3 = Eigen::Matrix3d::Identity();
+  Eigen::MatrixXd identity6 = Eigen::MatrixXd::Identity(6, 6);
+  Eigen::VectorXd eigenvector = Eigen::VectorXd::Random(6);
+  // case LINEAR
```
I was actually thinking about that. Makes total sense.

buschbapti commented:
@eeberhard I have addressed your comments thank for those. Concerning the test case with different damping, is it sufficient to have different damping eigenvector values (e1 and e2)? I am not sure what is expected to be generated.

domire8 reviewed (approved):
All good from my side

eeberhard commented:
> @eeberhard I have addressed your comments thank for those. Concerning the test case with different damping, is it sufficient to have different damping eigenvector values (e1 and e2)? I am not sure what is expected to be generated.

I think the extra test case would be very useful, since the point of tests is to eliminate the concept of "not sure what to expect". I suggest you initialise the controller with damping values such as <10, 1, 1, 1, 1, 1> or any other values, and then test the linear case where you give a desired velocity of <1, 0, 0> and a real velocity of <0, 0, 0>. In that case you expect a control wrench of 10 * 1 in just the X axis. Then you give a desired velocity of <1, 0, 0> and a real velocity of <1, 1, 0>, and I believe you would get a control wrench of 1 * 1 in Y and 0 in X. Something along those lines, obviously the numbers and axes tested can be different.

Essentially you would not just be testing the compute_damping, but the compute_command step itself.

buschbapti commented:
Good point I will implement that now

buschbapti commented:
Test has been added @eeberhard I have tested the differences in the equation as we discussed and I get the same results. I will close this PR. Later we can compare with already implemented PassiveDS and make some adjustments if needed.
